saw-core/src/SAWCore/SCTypeCheck.hs

```diff
@@ -88,40 +84,30 @@ type TCState = Map TermIndex Term
 data TCEnv =
   TCEnv
   { tcSharedContext :: SharedContext -- ^ the SAW context
-  , tcModName :: Maybe ModuleName    -- ^ the current module name, if any
   , tcCtx :: [(LocalName, Term)]     -- ^ the mapping of names to de Bruijn bound variables
-  , tcCtxEC :: Map LocalName (ExtCns Term) -- ^ the mapping of names to named variables
   }
 
 -- | The monad for type checking and inference, which:
 --
--- * Maintains a 'SharedContext', the name of the current module, and a variable
--- context, where the latter assigns types to the deBruijn indices in scope;
+-- * Maintains a 'SharedContext' and a variable context, where the
+--   latter assigns types to the deBruijn indices in scope;
 --
 -- * Memoizes the most general type inferred for each expression; AND
 --
 -- * Can throw 'TCError's
-type TCM = ReaderT TCEnv (StateT TCState (ExceptT TCError IO))
+newtype TCM a = TCM (ReaderT TCEnv (StateT TCState (ExceptT TCError IO)) a)
+  deriving (Functor, Applicative, Monad, MonadFail, MonadIO)
```

Comment:
It might be worth deriving `MonadReader TCEnv`, `MonadState TCState`, and `MonadError TCError` instances as well. That way, you wouldn't need to explicitly wrap/unwrap newtypes as much in the functions below. (For instance, you could implement `askCtx` as `askCtx = asks tcCtx` instead of `askCtx = TCM (asks tcCtx)`.)